autogen/coding/jupyter_code_executor.py

```diff
@@ -0,0 +1,221 @@
+import base64
+import json
+import os
+from pathlib import Path
+import re
+import uuid
+from typing import Any, ClassVar, List, Union
+
+from pydantic import Field
+
+
+from ..agentchat.agent import LLMAgent
+from .base import CodeBlock, CodeExecutor, CodeExtractor, CodeResult, IPythonCodeResult
+from .markdown_code_extractor import MarkdownCodeExtractor
+from .jupyter import JupyterConnectable, JupyterConnectionInfo, LocalJupyterServer, JupyterClient
+
+__all__ = ("JupyterCodeExecutor", "LocalJupyterCodeExecutor")
+
+
+class JupyterCodeExecutor(CodeExecutor):
+    """(Experimental) A code executor class that executes code statefully using an embedded
+    IPython kernel managed by this class.
+
+    **This will execute LLM generated code on the local machine.**
+
+    Each execution is stateful and can access variables created from previous
+    executions in the same session. The kernel must be installed before using
+    this class. The kernel can be installed using the following command:
+    `python -m ipykernel install --user --name {kernel_name}`
+    where `kernel_name` is the name of the kernel to install.
+
+    Args:
+        timeout (int): The timeout for code execution, by default 60.
+        kernel_name (str): The kernel name to use. Make sure it is installed.
+            By default, it is "python3".
+        output_dir (str): The directory to save output files, by default ".".
+        system_message_update (str): The system message update to add to the
+            agent that produces code. By default it is
+            `IPythonCodeExecutor.DEFAULT_SYSTEM_MESSAGE_UPDATE`.
```

Comment:
```suggestion
            `JupyterCodeExecutor.DEFAULT_SYSTEM_MESSAGE_UPDATE`.
```

Reply:
I think in the future we need to decouple the system message update from this code executor. If more languages are to be added this default system message is going to be outdated.